File: lib/include/type.lua
-- Copyright 2023 The Yock Authors. All rights reserved.
-- Use of this source code is governed by a MIT-style
-- license that can be found in the LICENSE file.

---@meta _

-----@alias err string | nil

---@class err: string

---@alias starType
---| BooleanType
---| StringType
---| StringArrayType

---@class BooleanType
local BooleanType = {}

---@param b boolean
---@return BooleanType
function Boolean(b)
    return {}
end

---@return userdata
function BooleanType:Ptr()
end

---@return boolean
function BooleanType:Var()
    return false
end

---@class StringType
local StringType = {}

---@param str string
---@return StringType
function String(str)
    return {}
end

---@return userdata
function StringType:Ptr()
end

---@return string
function StringType:Var()
    return ""
end

---@class StringArrayType
local StringArrayType = {}

---@vararg string
---@return StringArrayType
function StringArray(...)
    return {}
end

---@return userdata
function StringArrayType:Ptr()
end

---@return table
function StringArrayType:Var()
    return {}
end

File: util/type.go
// Copyright 2023 The Yock Authors. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package util

type Comparable interface {
	~int | ~int8 | ~int16 | ~int32 | ~int64 | ~uint | ~uint8 |
		~uint16 | ~uint32 | ~uint64 | ~uintptr | ~float32 | ~float64 | ~string
}
